lib/puppet/file_system/windows.rb

```diff
@@ -3,6 +3,27 @@
 
 class Puppet::FileSystem::Windows < Puppet::FileSystem::Posix
 
+  def expand_path(path)
+    string_path = ::File.expand_path(path_string(path))
+    # if no tildes, nothing to expand, return original string
+    # WARN: this will return a / on Windows if passed a /
+    return string_path if !string_path.index('~')
+
+    begin
+      string_path = Puppet::Util::Windows::File.get_long_pathname(string_path)
+      # TODO: ignore file not found errors and pass out original string
+      # to preserve the regular expand_path behavior
+      # no need to do existence check up front as GetLongPathName implies that check is performed
+      # and it should be the exception that files aren't actually present
+    rescue Puppet::Util::Windows::Error => e
+      raise if (e.code != Puppet::Util::Windows::File::ERROR_FILE_NOT_FOUND &&
+        e.code != Puppet::Util::Windows::File::ERROR_PATH_NOT_FOUND)
+      puts "#{e.code} error on with GetLongPathName and #{path}\n"
```

Comment:
Whoops - that `puts` statement was only used during debugging - I have to remove that and the above `TODO`